torchao/prototype/awq/api.py

```diff
@@ -0,0 +1,109 @@
+
+    # AQT config
+    mapping_type = MappingType.ASYMMETRIC
+    block_size = (1, group_size)
+    quant_min = 0
+    quant_max = 255 if quant_dtype == torch.uint8 else 2 ** _DTYPE_TO_BIT_WIDTH[quant_dtype] - 1 
+    eps = torch.finfo(torch.float32).eps
+    preserve_zero = True
+    zero_point_dtype = torch.int64
+    zero_point_domain = ZeroPointDomain.INT
+    assert quant_dtype in _DTYPE_TO_BIT_WIDTH or quant_dtype == torch.uint8, "Invalid quant_dtype. Please use torch.uint1 .. torch.uint8"
+
+    def replace_with_observer(layer):
+        # creates observer and replaces linear layers with observed linear layers
+        observer = AWQObserver(
+            layer.weight,
+            layer.bias, 
+            block_size, 
+            mapping_type,
+            quant_dtype, 
+            scale_search_space_size,
+            preserve_zero = preserve_zero,
+            zero_point_domain = zero_point_domain,
+            zero_point_dtype = zero_point_dtype,
+            quant_min=quant_min,
+            quant_max = quant_max,
+            eps = eps)
+        return ObservedLinear.from_float(layer, observer)
+    _replace_with_custom_fn_if_matches_filter(model, replace_with_observer, _is_linear)
+
+def _observed_linear_subclass_inserter(constructor):
+    """
+    Replaces unquantized observed linear instances with quantized linear instances.
+
+    Args:
+        constructor: the function which applies quantization to the observed linear layer
+    """
+    def insert_subclass(observed_linear):
+        # creates the new linear layer using constructor
+        linear = torch.nn.Linear(observed_linear.in_features, observed_linear.out_features, observed_linear.bias!=None, device=observed_linear.weight.device, dtype=observed_linear.weight.dtype)
+        linear.weight = torch.nn.Parameter(constructor(observed_linear), requires_grad=False)
+        linear.bias = observed_linear.bias
+        return linear
+
+    return insert_subclass
+
+def awq_uintx(quant_dtype: torch.dtype = torch.uint4, group_size: int = 128):
+    """
+    Quantizes linear layers when passed into quantize_()
+
+    Args:
+        quant_dtype: The data type of the quantized weights. Currently only torch.uint4 is intended to be used but can be used with torch.uint1 -> torch.uint8
+        group_size: Quantization granularity. Use -1 for channel wise quantization
+    """
+    
+    assert quant_dtype in _DTYPE_TO_BIT_WIDTH or quant_dtype == torch.uint8, "Invalid quant_dtype. Please use torch.uint1 .. torch.uint8"
+    def weight_quant_func(observed_linear):
+        # weight quantization
+        equalization_scale = observed_linear.act_obs.calculate_qparams()
+        # AQT config
+        target_dtype = torch.uint8
+        mapping_type = MappingType.ASYMMETRIC
+        block_size = (1, group_size)
+        quant_min = 0
+        quant_max = 255 if quant_dtype == torch.uint8 else 2 ** _DTYPE_TO_BIT_WIDTH[quant_dtype] - 1 
```

Comment:
you can use https://github.com/pytorch/ao/blob/144445a7a8e988059555421caafa17e0c1678053/torchao/quantization/quant_primitives.py#L75 I think